src/Core/Content/Mail/Message/SendMailHandler.php

```diff
@@ -0,0 +1,69 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Core\Content\Mail\Message;
+
+use League\Flysystem\FilesystemOperator;
+use Shopware\Core\Framework\Log\Package;
+use Symfony\Component\Mailer\Exception\TransportExceptionInterface;
+use Symfony\Component\Mailer\Transport\TransportInterface;
+use Symfony\Component\Messenger\Attribute\AsMessageHandler;
+use Symfony\Component\Messenger\Exception\UnrecoverableMessageHandlingException;
+use Symfony\Component\Mime\Email;
+
+/**
+ * @internal
+ */
+#[AsMessageHandler(handles: SendMailMessage::class)]
+#[Package('services-settings')]
+final class SendMailHandler
+{
+    /**
+     * @internal
+     */
+    public function __construct(
+        private readonly TransportInterface $transport,
+        private readonly FilesystemOperator $filesystem,
+    ) {
+    }
+
+    /**
+     * @throws TransportExceptionInterface
+     * @throws UnrecoverableMessageHandlingException
+     */
+    public function __invoke(SendMailMessage $message): void
+    {
+        $mailDataPath = $message->getMailDataPath();
+
+        try {
+            $mailData = $this->filesystem->read($mailDataPath);
+        } catch (\Throwable $e) {
+            throw new UnrecoverableMessageHandlingException(\sprintf('An error occurred while reading the mail data file "%s".', $mailDataPath), 0, $e);
```

Comment:
Can you distinguish between file does not exist and other errors. File does not exist would be an unrecoverable, others e.g. caused by temporary network issues would make sense to retry them.

Reply:
As far as I can see in the FilesystemAdapters of e.g. Local or S3, they all just catch all errors and throw a UnableToReadFile themselves. So we will always get the same error. Unless we explicitly check if the file exists beforehand, we have no way of knowing what the reason for the error is.
The file could be temporarily or permanently gone.

Reply:
@keulinho @shyim What do you think about this & about https://github.com/shopware/shopware/pull/5490#discussion_r1841976012?

Reply:
Yes, they just throw UnableToReadFile exceptions.

Reply:
But I would make the message more clear by adding the information, that the mail has not been sent.

Reply:
What I would do, is place the fileExists call inside the catch exception block. In that case, the 2 calls to the storage would be made only in edge cases where we need to be sure that the error is recoverable or not

Reply:
@Gaitholabi I added the check and more tests.

Reply:
hey, a quick update, we're still looking at this and looking to get it done soon. thanks!